airflow/utils/log/file_task_handler.py

```diff
@@ -0,0 +1,186 @@
+        elif try_number > current_try or try_number < 0:
+            logs = ['Error fetching the log. Try number {} is invalid.'.format(try_number)]
+            return logs
+        else:
+            try_numbers = [try_number]
+
+        logs = [''] * len(try_numbers)
+        for i, try_number in enumerate(try_numbers):
+            logs[i] += self._read(task_instance, try_number)
+
+        return logs
+
+    def _get_log_relative_dir(self, task_instance):
+        """
+        Get relative log file directory.
+        :param task_instance: task instance object
+        """
+        return os.path.join(task_instance.dag_id, task_instance.task_id,
+                            task_instance.execution_date.isoformat())
+
+    def _get_log_relative_path(self, task_instance, try_number=None):
+        """
+        Get relative log file path.
+        :param task_instance: task_instance object
+        :param try_number: optional task instance try_number to
+        override task instance try_number
+        """
+        directory = self._get_log_relative_dir(task_instance)
+        try_number = task_instance.try_number if try_number is None else try_number
+        # For reporting purposes and keeping logs consistent with web UI
+        # display, we write and report based on 1-indexed, not 0-indexed lists
+        filename = "{}.log".format(try_number + 1)
+        return os.path.join(directory, filename)
+
+    def _init_file(self, task_instance):
+        """
+        Create log directory and give it correct permissions.
+        :param task_instance: task instance object
+        :return relative log path of the given task instance
+        """
+        # To handle log writing when tasks are impersonated, the log files need to
+        # be writable by the user that runs the Airflow command and the user
+        # that is impersonated. This is mainly to handle corner cases with the
+        # SubDagOperator. When the SubDagOperator is run, all of the operators
+        # run under the impersonated user and create appropriate log files
+        # as the impersonated user. However, if the user manually runs tasks
+        # of the SubDagOperator through the UI, then the log files are created
+        # by the user that runs the Airflow command. For example, the Airflow
+        # run command may be run by the `airflow_sudoable` user, but the Airflow
+        # tasks may be run by the `airflow` user. If the log files are not
+        # writable by both users, then it's possible that re-running a task
+        # via the UI (or vice versa) results in a permission error as the task
+        # tries to write to a log file created by the other user.
+        relative_log_dir = self._get_log_relative_dir(task_instance)
+        directory = os.path.join(self.local_base, relative_log_dir)
+        # Create the log file and give it group writable permissions
+        # TODO(aoen): Make log dirs and logs globally readable for now since the SubDag
```

Comment:
@aoenj just thinking out loud: if task_instances would log to stdout and we would let the executor setup logging before it does "sudo" we would not need to set 666 on this file and don't have to deal with permissions for logging at all?